lib/Doctrine/DBAL/Connection.php

```diff
@@ -1172,15 +1172,17 @@ public function transactional(Closure $func)
         $this->beginTransaction();
         try {
             $res = $func($this);
-            $this->commit();
-            return $res;
         } catch (Exception $e) {
             $this->rollBack();
             throw $e;
         } catch (Throwable $e) {
             $this->rollBack();
             throw $e;
         }
+
+        $this->commit();
```

Comment:
Shouldn't a crash on `commit()` lead to a rollback as well?

Reply:
That is precisely the issue here - when you call `commit()` and it fails, then the transaction is no more :) There is nothing to rollback ... so this PR is all about not calling the rollback if the failure occurs in the `commit()`

Reply:
> then the transaction is no more

Ok, this is my misconception then. Are we sure that this holds consistently for all platforms that we support?

Reply:
AFAIK all other platforms throw the exceptions before `commit()` (they do not support constraint deferring), so we will always call rollback in those cases. But that might not be true - I'm not familiar enough with all of them to be perfectly honest. 

Anyway, shouldn't the existing tests cover that?

Reply:
> shouldn't the existing tests cover that?

If they run on all platforms, yes.

Reply:
I looked at the functional tests of `Connection` class and it looks like there are no tests covering constraints violations, only basic nesting level checks and whatnot :/ I'll try to write some on Monday.

Reply:
@Ocramius I improved the ConnectionTest functional test https://github.com/doctrine/dbal/pull/3425 . Could that be enough? If so, I can rebase this branch onto it...

Reply:
I'll have to defer that decision to @deeky666 and @morozov

Reply:
What about a case where the Exception happens during an open transaction? Shouldn't this improvement be marked as a BC break, when there are no rollbacks anymore?

Reply:
@SenseException I'm sorry, I don't understand. I don't think that I changed the behavior for "common" transactions in any way; please provide an example code if you think so - thanks.

Reply:
You fixed your issue "Rollback called when outside of a transaction" in this method by moving the `commit()` call out of the try-catch. Before your change an exception thrown in `commit()` would call a rollback in the catch, but now this isn't the case anymore. What if `commit()` throws an exception inside a transaction and a rollback is needed?

Reply:
Thanks, yeah, I understand now.

> What if commit() throws an exception inside a transaction and a rollback is needed?

Then that case wouldn't be handled. The thing is that I was not able to come up with an example where `COMMIT` would throw an exception and also where would need to call `ROLLBACK` afterwards. I asked all devs in my company and no one was able to think of a situation like that. I also asked one guy who knows basically everything around MySQL and he was not able to think of such example (even tried poking master/slave and cluster setups).

But I guess unless someone checks all source codes/documentations of all supported databases we can't be 100 % sure. All I can do is to deploy those changes to our production systems (Postgres and MySQL) and report back if we experienced anything unusual...

Or can you think of a better solution to this problem?